### PR TITLE
Add ability to open user profile links with username instead of user ID

### PR DESCRIPTION
**Pull request by rednir**
Currently links such as https://osu.ppy.sh/u/username are ignored when trying to open in-game, as only user IDs are allowed.

This adds a constructor to `GetUserRequest` which takes in a username instead of user ID, and adds an overload to `ShowUser()`

### CODE REVIEW

File: osu.Game/Overlays/UserProfileOverlay.cs [hunk 1]
```diff
@@ -116,7 +118,7 @@ public void ShowUser(User user, bool fetchOnline = true)
 
             if (fetchOnline)
             {
-                userReq = new GetUserRequest(user.Id);
+                userReq = user.Username != null ? new GetUserRequest(user.Username) : new GetUserRequest(user.Id);
```
Comment (smoogipoo) [inline osu.Game/Overlays/UserProfileOverlay.cs:121]: Should probably be the other way around - only use username if `Id <= 1` because `Id` should always be more accurate.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
For reference, how do you get to that overlay with only a username currently?

File: osu.Game/Overlays/UserProfileOverlay.cs (hunk 1 above)
Comment (rednir) [reply osu.Game/Overlays/UserProfileOverlay.cs]: Sure, did `Id > 1 ? new GetUserRequest(user.Id)` instead because I think it's slightly more readable

**Comment by rednir** [comment]
Uh... I didn't think you could? I thought the overlay always got its data with a user ID or with the `User` object fetched beforehand.

**Comment by frenzibyte** [comment]
More like what do you need this PR for? it doesn't seem like `OsuGameBase.ShowUser(string username)` is ever called from anywhere? is this missing further implementation or intended to be a follow-up PR?

**Comment by smoogipoo** [comment]
@frenzibyte that method has a `User` overload.

**Comment by rednir** [comment]
It's just something I noticed when trying to open links like https://osu.ppy.sh/u/username in the chat, the fact that the API allows that kind of link but the lazer client does not. I was typing out the link in chat to open the profile of someone from username while in-game. I suppose it's a bit niche.

>is this missing further implementation or intended to be a follow-up PR?

I didn't intend to open a follow up PR, but thinking about it now, I'm not sure there is an easy way for the user to open the profile of someone from their username alone. Maybe that's something to consider.

**Comment by frenzibyte** [comment]
Nevermind, I didn't notice that change in `HandleLink` (I should've checked out the branch instead).

**Comment by smoogipoo** [comment]
One place where this could have an effect is clicking on the author of https://osu.ppy.sh/beatmapsets/41823#osu/131891. Doesn't seem to be working right now because it gets `Username = "0"` anyway :stuck_out_tongue:

**Comment by rednir** [comment]
image
Am I missing something because this looks like its working normal to me? Looks like correct `Username` and `Id`

**Comment by smoogipoo** [comment]
Yes, you have to go one level deeper:
image

**Comment by frenzibyte** [comment]
Yep, https://github.com/ppy/osu/issues/5966.

File: osu.Game/OsuGame.cs [hunk 2]
```diff
@@ -329,8 +329,7 @@ private void load()
                     break;
 
                 case LinkAction.OpenUserProfile:
-                    if (int.TryParse(link.Argument, out int userId))
-                        ShowUser(userId);
+                    ShowUser(link.Argument);
```
Comment (frenzibyte) [inline osu.Game/OsuGame.cs:332]: While it works either way, I believe the ID overload should be used when it's supposed to be, rather than making all cases go for the "username" overload to eventually end up having the same target URL during request.
```suggestion
                    if (int.TryParse(link.Argument, out var userId))
                        ShowUser(userId);
                    else
                        ShowUser(link.Argument);

```

File: osu.Game/Online/API/Requests/GetUserRequest.cs [hunk 3]
```diff
@@ -8,15 +8,25 @@ namespace osu.Game.Online.API.Requests
...
+        }
+
+        public GetUserRequest(string username = null, RulesetInfo ruleset = null)
+        {
+            this.userIdentifier = username;
             Ruleset = ruleset;
         }
 
-        protected override string Target => userId.HasValue ? $@"users/{userId}/{Ruleset?.ShortName}" : $@"me/{Ruleset?.ShortName}";
+        protected override string Target => (userIdentifier != null) ? $@"users/{userIdentifier}/{Ruleset?.ShortName}" : $@"me/{Ruleset?.ShortName}";
```
Comment (frenzibyte) [inline osu.Game/Online/API/Requests/GetUserRequest.cs:30]:
```suggestion
        protected override string Target => userIdentifier != null ? $@"users/{userIdentifier}/{Ruleset?.ShortName}" : $@"me/{Ruleset?.ShortName}";
```

**Review by frenzibyte — CHANGES_REQUESTED** [review]
In addition to the above (which seems feasible to have I believe).

**Comment by rednir** [comment]
Is it sensible to sneak in a fix for #5966 in this PR? It seems relatively more involved (mainly the fetching avatar bit)

**Comment by peppy** [comment]
I would highly advise we never rely on username lookups for anything which can avoid it. If we only have a username due to the source only providing that, we should convert it to an ID at import time, not use the username at runtime.

**Comment by rednir** [comment]
I'm not very confident in the way I have implemented this but I'd like some input on the above commit, just not sure if I should be making an API request for the entire user. It does get the avatar on scores imported from stable working correctly though AFAIK.

File: osu.Game/OsuGame.cs [hunk 4]
```diff
@@ -329,8 +329,11 @@ private void load()
                     break;
 
                 case LinkAction.OpenUserProfile:
-                    if (int.TryParse(link.Argument, out int userId))
+                    if (int.TryParse(link.Argument, out var userId))
```
Comment (peppy) [inline osu.Game/OsuGame.cs:332]: why was this changed

File: osu.Game/Scoring/ScoreManager.cs [hunk 5]
```diff
@@ -72,8 +76,31 @@ protected override ScoreInfo CreateModel(ArchiveReader archive)
             }
         }
 
+        private readonly Dictionary<string, User> previouslyLookedUpUsernames = new Dictionary<string, User>();
```
Comment (peppy) [inline osu.Game/Scoring/ScoreManager.cs:79]: this is permanent memory usage.

you want to be using https://github.com/ppy/osu/blob/master/osu.Game/Database/UserLookupCache.cs if anything.

File: osu.Game/Scoring/ScoreManager.cs [hunk 6]
```diff
@@ -72,8 +76,31 @@ protected override ScoreInfo CreateModel(ArchiveReader archive)
...
+
         protected override Task Populate(ScoreInfo model, ArchiveReader archive, CancellationToken cancellationToken = default)
-            => Task.CompletedTask;
+        {
+            // These scores only provide the user's username but we need the user's ID too.
+            if (model.UserID <= 1 && model.UserString != null)
```
Comment (peppy) [inline osu.Game/Scoring/ScoreManager.cs:84]: there's zero chance this is working. you're doing an async request and not blocking. work by exxample there's likely a simialr thing in beatmap import or oterhwise.

File: osu.Game/Online/API/Requests/GetUserRequest.cs [hunk 7]
```diff
@@ -8,15 +8,25 @@ namespace osu.Game.Online.API.Requests
...
     public class GetUserRequest : APIRequest<User>
     {
-        private readonly long? userId;
+        private readonly string userIdentifier;
         public readonly RulesetInfo Ruleset;
 
+        public GetUserRequest()
+        {
+        }
+
```
Comment (peppy) [inline osu.Game/Online/API/Requests/GetUserRequest.cs:17]: needs xmldoc at very least

File: osu.Game/Online/Chat/ChannelManager.cs [hunk 8]
```diff
@@ -256,8 +256,21 @@ public void PostCommand(string text, Channel target = null)
                     JoinChannel(channel);
                     break;
 
+                case "chat":
```
Comment (peppy) [inline osu.Game/Online/Chat/ChannelManager.cs:259]: you're really overloading this PR here. please choose one feature to implement and do that first.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
This PR is adding
- score user fix
- chat commands
- a new *untested and unused* flow to profile overlay (already mentioend https://github.com/ppy/osu/pull/14571#issuecomment-907943315 but still an issue)

File: osu.Game/Scoring/ScoreManager.cs (hunk 5 above)
Comment (rednir) [reply osu.Game/Scoring/ScoreManager.cs]: `UserLookupCache` only accepts user IDs right now. It could be modified to take in a string instead of int, but it uses a `GetUsersRequest` to fetch data from online, and that API call only accepts user IDs as a parameter.

Made a new class with similar implementation to `UserLookupCache` (meaning I copied half of it) that gets an ID from username, it uses `GetUserRequest` for each user instead of doing them in a batch. I feel like I'm missing an more sensible way though.

**Comment by rednir** [comment]
Oops sorry, pushed these changes without realizing how much worse I made this PR in terms of size... I'll revert and leave only the new profile overlay flow for this PR.

**Comment by peppy** [comment]
I feel like i'm braindead so let me ask again: where is this used? how am i supposed to test this? This flow should *not exist unless utmost required*. As already mentioned, I *don't think it ever should be*.

**Comment by rednir** [comment]
I thought my previous comment answered that?
>It's just something I noticed when trying to open links like https://osu.ppy.sh/u/username in the chat, the fact that the API allows that kind of link but the lazer client does not. I was typing out the link in chat to open the profile of someone from username while in-game. I suppose it's a bit niche.

**Comment by peppy** [comment]
How are you finding such links in chat? We use user ID in almost all places.

**Comment by rednir** [comment]
`I was typing out the link in chat to open the profile of someone from username while in-game. I suppose it's a bit niche.`

**Comment by peppy** [comment]
Really unsure this should be supported since we never use username based links anywhere, but for now it's a whatever I guess.

File: osu.Game/Online/API/Requests/GetUserRequest.cs [hunk 9]
```diff
@@ -8,15 +8,38 @@ namespace osu.Game.Online.API.Requests
 {
     public class GetUserRequest : APIRequest<User>
     {
-        private readonly long? userId;
+        private readonly string userIdentifier;
```
Comment (peppy) [inline osu.Game/Online/API/Requests/GetUserRequest.cs:11]: this whole request is probably too overloaded at this point but eh
